firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +314,259 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
 
         self.uc = firedrake.Function(Vc)
         self.uf = firedrake.Function(Vf)
-        self.prolong_kernel = self.prolongation_transfer_kernel_action(Vf, self.uc)
 
+        self.mesh = Vf.mesh()
+        self.weight = self.multiplicity(Vf)
+        with self.weight.dat.vec as w:
+            w.reciprocal()
+
+        tf, _, _ = tensor_product_space_query(Vf)
+        tc, _, _ = tensor_product_space_query(Vc)
+        if tf and tc:
+            self.make_blas_kernels(Vf, Vc)
+        else:
+            self.make_kernels(Vf, Vc)
+        return
+
+    def make_kernels(self, Vf, Vc):
+        """
+        Interpolation and restriction kernels between arbitrary elements.
+
+        This is temporary while we wait for structure-preserving tfsc kernels.
+        """
+        self.prolong_kernel = self.prolongation_transfer_kernel_action(Vf, self.uc)
         matrix_kernel = self.prolongation_transfer_kernel_action(Vf, firedrake.TestFunction(Vc))
+        # The way we transpose the prolongation kernel is suboptimal.
+        # A local matrix is generated each time the kernel is executed.
```

Comment:
If you care, but I think one should just not spend too much effort on this, you could evaluate on a single element and then hard code those numbers.

Reply:
I wouldn't like to spend time on that.